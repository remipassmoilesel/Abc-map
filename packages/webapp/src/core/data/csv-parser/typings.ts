/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

export declare type CsvRow = { [k: string]: string | number | undefined };

export const UnknownLineNumber = -1;

export class CsvParsingError extends Error {
  constructor(message: string, public line: number) {
    super(message);
  }
}

export function isCsvParsingError(err: unknown): err is CsvParsingError {
  return !!err && typeof err === 'object' && 'line' in err && typeof err.line !== 'undefined';
}
