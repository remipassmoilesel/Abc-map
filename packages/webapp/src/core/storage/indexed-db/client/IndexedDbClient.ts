/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Logger } from '@abc-map/shared';
import isEqual from 'lodash/isEqual';
import { ObjectStore } from './ObjectStore';

export const logger = Logger.get('IndexedDbClient.ts', 'warn');

export interface KvPair<T = unknown> {
  key: IDBValidKey;
  value: T;
}

export function toKvPair<T extends { id: IDBValidKey }>(objs: T[]): KvPair<T>[] {
  return objs.map((obj) => ({ key: obj.id, value: obj }));
}

export type UpgradeHandler = (db: IDBDatabase, req: IDBOpenDBRequest, ev: IDBVersionChangeEvent) => void;
export type CursorHandler<T> = (value: T | null, cursor: IDBCursorWithValue | null) => void;
export type CursorErrorHandler = (err: unknown) => void;

interface IndexedDbEntry<T> {
  key: IDBValidKey;
  value: T | undefined;
}

// See: https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB
export class IndexedDbClient {
  private database?: IDBDatabase;

  public async connect(databaseName: string, version = 1, onUpgradeNeeded: UpgradeHandler): Promise<IDBDatabase> {
    if (this.database) {
      throw new Error('Already connected');
    }

    return new Promise((resolve, reject) => {
      const request = window.indexedDB.open(databaseName, version);

      request.onupgradeneeded = (ev) => onUpgradeNeeded(request.result, request, ev);

      request.onerror = (event) => {
        logger.error(`Database error:`, (event.target as IDBRequest)?.error);
        reject(new Error('Unexpected error'));
      };

      request.onsuccess = () => {
        this.database = request.result;
        resolve(request.result);
      };
    });
  }

  public disconnect(): void {
    if (!this.database) {
      throw new Error('Not connected');
    }

    this.database.close();
  }

  private getDatabase(): IDBDatabase {
    if (!this.database) {
      throw new Error('Not connected');
    }
    return this.database;
  }

  public async exists(storeName: string, key: string): Promise<boolean> {
    const db = this.getDatabase();
    const store = db.transaction(storeName, 'readonly').objectStore(storeName);

    return this.requestPromise(store.getKey(key)).then((res) => !!res.result);
  }

  public async put<T>(storeName: string, key: string, value: T): Promise<void> {
    const db = this.getDatabase();
    const transaction = db.transaction(storeName, 'readwrite');
    const store = transaction.objectStore(storeName);

    store.put(value, key);

    return this.transactionPromise(transaction).then(() => undefined);
  }

  public async putAll<T>(storeName: string, pairs: KvPair<T>[]): Promise<IDBTransaction> {
    const db = this.getDatabase();
    const transaction = db.transaction(storeName, 'readwrite', { durability: 'relaxed' });
    const store = transaction.objectStore(storeName);

    pairs.forEach(({ key, value }) => store.put(value, key));
    transaction.commit();

    return this.transactionPromise(transaction);
  }

  public async deleteAll(storeName: string, keys: (IDBValidKey | IDBKeyRange)[]): Promise<IDBTransaction> {
    const db = this.getDatabase();
    const transaction = db.transaction(storeName, 'readwrite', { durability: 'relaxed' });
    const store = transaction.objectStore(storeName);

    keys.forEach((key) => store.delete(key));
    transaction.commit();

    return this.transactionPromise(transaction);
  }

  public async get<T>(storeName: string, key: string): Promise<T | undefined> {
    const db = this.getDatabase();
    const store = db.transaction(storeName, 'readonly').objectStore(storeName);

    return this.requestPromise(store.get(key)).then((res) => res.result);
  }

  public async getAllByKeys<T>(storeName: string, keys: IDBValidKey[]): Promise<T[]> {
    if (!keys) {
      throw new Error('"keys" must be defined');
    }
    if (!keys.length) {
      return [];
    }

    const result: IndexedDbEntry<T>[] = [];

    return new Promise((resolve, reject) => {
      this.openCursor<T>(
        storeName,
        (value, cursor) => {
          const hasMore = value != null && typeof value !== 'undefined' && typeof cursor?.key !== 'undefined';
          if (hasMore) {
            const match = keys.find((key) => isEqual(key, cursor.key));
            if (match) {
              result.push({ key: cursor.key, value });
            }

            cursor.continue();
          } else {
            // We try to reproduce order from input
            const sorted = keys
              .map((inputKey) => result.find(({ key }) => key === inputKey))
              .map((entry) => entry?.value)
              .filter((value): value is T => !!value);

            resolve(sorted);
          }
        },
        (err) => {
          reject(err);
        }
      );
    });
  }

  public async getAll<T>(storeName: string): Promise<{ key: IDBValidKey; value: T }[]> {
    const result: { key: IDBValidKey; value: T }[] = [];

    return new Promise((resolve, reject) => {
      this.openCursor<T>(
        storeName,
        (value, cursor) => {
          const hasMore = value != null && typeof value !== 'undefined' && typeof cursor?.key !== 'undefined';
          if (hasMore) {
            result.push({ key: cursor?.key, value: value });
            cursor.continue();
          } else {
            resolve(result);
          }
        },
        (err) => {
          reject(err);
        }
      );
    });
  }

  public openCursor<T>(storeName: string, handler: CursorHandler<T>, errorHandler: CursorErrorHandler): void {
    try {
      const db = this.getDatabase();
      const store = db.transaction(storeName, 'readonly').objectStore(storeName);
      const cursorReq = store.openCursor();

      cursorReq.onsuccess = () => {
        handler(cursorReq.result?.value ?? null, cursorReq.result);
      };

      cursorReq.onerror = () => {
        errorHandler(cursorReq.error);
      };
    } catch (err) {
      errorHandler(err);
    }
  }

  public openIndexCursor<T>(storeName: string, indexName: string, keyRange: IDBKeyRange, handler: CursorHandler<T>, errorHandler: CursorErrorHandler): void {
    try {
      const db = this.getDatabase();
      const store = db.transaction(storeName, 'readonly').objectStore(storeName);
      const index = store.index(indexName);
      const cursorReq = index.openCursor(keyRange);

      cursorReq.onsuccess = () => {
        handler(cursorReq.result?.value ?? null, cursorReq.result);
      };

      cursorReq.onerror = () => {
        errorHandler(cursorReq.error);
      };
    } catch (err) {
      errorHandler(err);
    }
  }

  public async clearDatabase() {
    const protectedStores: string[] = [ObjectStore.Migrations];
    const db = this.getDatabase();
    for (const storeName of db.objectStoreNames) {
      if (protectedStores.includes(storeName)) {
        continue;
      }

      try {
        const transaction = db.transaction(storeName, 'readwrite', { durability: 'relaxed' });
        const store = transaction.objectStore(storeName);
        store.clear();
        transaction.commit();

        await this.transactionPromise(transaction);
      } catch (err) {
        logger.error('Store deletion error', err);
      }
    }
  }

  private requestPromise<T>(request: IDBRequest<T>): Promise<IDBRequest<T>> {
    return new Promise((resolve, reject) => {
      request.addEventListener('error', () => reject(request.error || new Error('Unexpected error')));
      request.addEventListener('success', () => resolve(request));
    });
  }

  private transactionPromise(transaction: IDBTransaction): Promise<IDBTransaction> {
    return new Promise((resolve, reject) => {
      transaction.addEventListener('error', () => reject(transaction.error || new Error('Unexpected error')));
      transaction.addEventListener('complete', () => resolve(transaction));
    });
  }
}
