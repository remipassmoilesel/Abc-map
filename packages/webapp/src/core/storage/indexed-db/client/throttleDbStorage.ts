/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import throttle from 'lodash/throttle';

/**
 * This function allow to limit storage call.
 *
 * With a 5s value, it will ensure that storageCb will be called once then no more the 5 next seconds.
 *
 * @param storageCb
 * @param waitBetweenCallsMs
 */
export function throttleDbStorage(storageCb: (...args: any[]) => void, waitBetweenCallsMs = 5_000) {
  return throttle(storageCb, waitBetweenCallsMs, { leading: true, trailing: true });
}
