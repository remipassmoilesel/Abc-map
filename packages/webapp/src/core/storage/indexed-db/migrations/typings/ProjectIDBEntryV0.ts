/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcView } from '@abc-map/shared';

export interface ProjectIDBEntryV0 {
  version?: number;
  // Metadata: id, version, etc ...
  metadata: IDBProjectMetadataV0;
  // Main view for edition
  view: AbcView;
  layerIds: string[];
  // Layouts for static exports
  layoutIds: string[];
  // Views publicly accessible
  sharedViews: {
    // If true, map is fullscreen
    fullscreen: boolean;
    // If not fullscreen, these dimensions are used
    mapDimensions: { width: number; height: number };
    // List of views
    viewIds: string[];
  };
}

export interface IDBProjectMetadataV0 {
  id: string;
  version: string;
  name: string;
  public: boolean;
  // This field has not been removed in a previous migration
  containsCredentials?: boolean;
}
