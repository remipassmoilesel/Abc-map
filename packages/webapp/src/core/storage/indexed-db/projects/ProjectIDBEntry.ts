/**
 * Copyright © 2022 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcProjectMetadata, AbcView } from '@abc-map/shared';

export const CURRENT_VERSION = 3;

export interface ProjectIDBEntry {
  version: number;
  // Metadata: id, version, etc ...
  metadata: AbcProjectMetadata;
  // Main view for edition
  view: AbcView;
  layerIds: string[];
  // Layouts for PDF exports
  layouts: {
    layoutIds: string[];
    abcMapAttributionsEnabled: boolean;
  };
  // Views publicly accessible
  sharedViews: {
    // If true, map is fullscreen
    fullscreen: boolean;
    // If not fullscreen, these dimensions are used
    mapDimensions: { width: number; height: number };
    // List of views
    viewIds: string[];
  };
}
