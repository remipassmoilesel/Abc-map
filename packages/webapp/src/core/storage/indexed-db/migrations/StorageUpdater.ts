/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { Logger } from '@abc-map/shared';
import { getMigrations, MigrationsFactory } from './migrations';
import { IndexedDbClient } from '../client/IndexedDbClient';
import { StorageMigrationContext } from './IndexedDbMigration';
import { MigrationIDBStorage } from './MigrationIDBStorage';

export const logger = Logger.get('StorageUpdater.ts', 'debug');

export function disableStorageMigrationLogs() {
  logger.disable();
}

/**
 * This helper migrate Indexed DB data
 */
export class StorageUpdater {
  public static create() {
    return new StorageUpdater(getMigrations);
  }

  constructor(private migrations: MigrationsFactory) {}

  public async update(client: IndexedDbClient): Promise<void> {
    const migrations = this.migrations();
    const migrationStorage = new MigrationIDBStorage(client);
    const migrationsAlreadyDone = await migrationStorage.getAll();

    const context: StorageMigrationContext = { client };

    for (const migration of migrations) {
      const alreadyDone = migrationsAlreadyDone.find((migr) => migr.name === migration.getName());
      if (!alreadyDone) {
        logger.warn(`Executing IndexedDB schema migration: ${migration.getName()}`);
        await migration.process(context);
        await migrationStorage.put({ name: migration.getName(), when: Date.now() });
      }
    }
  }
}
