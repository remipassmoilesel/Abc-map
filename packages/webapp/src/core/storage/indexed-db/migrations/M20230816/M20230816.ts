/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { IndexedDbMigration, StorageMigrationContext } from '../IndexedDbMigration';
import { IDBProjectMetadataV0 } from '../typings/ProjectIDBEntryV0';
import { ObjectStore } from '../../client/ObjectStore';
import { ProjectIDBEntry } from '../../projects/ProjectIDBEntry';
import { Logger } from '@abc-map/shared';

const NEXT = 3;

const logger = Logger.get('M20230816.ts');

/**
 * This migration removes "containsCredentials" field
 */
export class M20230816 implements IndexedDbMigration {
  public getName(): string {
    return 'M20230816';
  }

  public async process(context: StorageMigrationContext): Promise<void> {
    const { client } = context;

    // Projects
    const projects = await client.getAll<ProjectIDBEntry>(ObjectStore.Projects);
    const migratedProjects = projects
      .filter((entry) => (entry.value.version ?? 0) < NEXT)
      .map<ProjectIDBEntry>((entry) => {
        const result: ProjectIDBEntry = {
          version: NEXT,
          metadata: {
            ...entry.value.metadata,
          },
          view: entry.value.view,
          layouts: entry.value.layouts,
          sharedViews: entry.value.sharedViews,
          layerIds: entry.value.layerIds,
        };

        // This field has not been removed in previous migration
        const metadata = result.metadata as Partial<IDBProjectMetadataV0>;
        if (typeof metadata.containsCredentials !== 'undefined') {
          delete metadata.containsCredentials;
        }

        return result;
      });

    if (migratedProjects.length) {
      await client.putAll<ProjectIDBEntry>(
        ObjectStore.Projects,
        migratedProjects.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );
    }
  }
}
