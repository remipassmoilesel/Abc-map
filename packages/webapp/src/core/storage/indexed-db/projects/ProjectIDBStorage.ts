/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { IndexedDbClient } from '../client/IndexedDbClient';
import { ObjectStore } from '../client/ObjectStore';
import { getMainDbClient } from '../main-database';
import { AbcLayout, AbcProjectManifest, AbcSharedView, LayerType, Logger } from '@abc-map/shared';
import { LayerIDBStorage } from '../layers/LayerIDBStorage';
import { ProjectIDBEntry } from './ProjectIDBEntry';
import { FeatureIDBStorage } from '../features/FeatureIDBStorage';
import { GenericReduxIDBStorage } from '../redux/GenericReduxIDBStorage';
import { LayoutIDBStorage } from '../layouts/LayoutIDBStorage';
import { SharedViewIDBStorage } from '../shared-views/SharedViewIDBStorage';

const logger = Logger.get('ProjectIDBStorage.tsx');

/**
 * This is the main storage, from which you can fetch projects and all associated objects.
 */
export class ProjectIDBStorage {
  public static create(): ProjectIDBStorage {
    return new ProjectIDBStorage(getMainDbClient, LayerIDBStorage.create(), LayoutIDBStorage, SharedViewIDBStorage, FeatureIDBStorage.create());
  }

  constructor(
    private getClient: () => IndexedDbClient,
    private layers: LayerIDBStorage,
    private layouts: GenericReduxIDBStorage<AbcLayout>,
    private sharedViews: GenericReduxIDBStorage<AbcSharedView>,
    private features: FeatureIDBStorage
  ) {}

  public async exists(projectId: string): Promise<boolean> {
    const client = this.getClient();
    return client.exists(ObjectStore.Projects, projectId);
  }

  public async put(project: ProjectIDBEntry): Promise<void> {
    const client = this.getClient();
    await client.putAll(ObjectStore.Projects, [{ key: project.metadata.id, value: project }]);
  }

  public async get(projectId: string): Promise<ProjectIDBEntry | undefined> {
    const client = this.getClient();
    return client.get(ObjectStore.Projects, projectId);
  }

  public async getManifest(projectId: string): Promise<AbcProjectManifest | undefined> {
    const storageEntry = await this.get(projectId);
    if (!storageEntry) {
      return;
    }

    const layers = await this.layers.getAll(storageEntry.layerIds);

    // We fetch features
    for (const layer of layers) {
      if (layer.type === LayerType.Vector) {
        layer.features = {
          ...layer.features,
          features: await this.features.getAllByLayerId(layer.metadata.id),
        };
      }
    }

    const layouts = await this.layouts.getAll(storageEntry.layouts.layoutIds);
    const sharedViews = await this.sharedViews.getAll(storageEntry.sharedViews.viewIds);

    return {
      metadata: storageEntry.metadata,
      view: storageEntry.view,
      layers,
      layouts: {
        list: layouts,
        abcMapAttributionsEnabled: storageEntry.layouts.abcMapAttributionsEnabled,
      },
      sharedViews: {
        fullscreen: storageEntry.sharedViews.fullscreen,
        mapDimensions: storageEntry.sharedViews.mapDimensions,
        list: sharedViews,
      },
    };
  }
}
