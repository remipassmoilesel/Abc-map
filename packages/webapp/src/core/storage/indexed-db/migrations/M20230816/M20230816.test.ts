/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { ProjectIDBEntryV0 } from '../typings/ProjectIDBEntryV0';
import { ObjectStore } from '../../client/ObjectStore';
import { ProjectIDBEntry } from '../../projects/ProjectIDBEntry';
import { IndexedDbClient } from '../../client/IndexedDbClient';
import { nanoid } from 'nanoid';
import { TestHelper } from '../../../../utils/test/TestHelper';
import { M20230816 } from './M20230816';
import { disableStorageMigrationLogs } from '../StorageUpdater';

disableStorageMigrationLogs();

describe('M20230816', () => {
  let dbName: string;
  let client: IndexedDbClient;
  let migration: M20230816;

  beforeEach(async () => {
    client = new IndexedDbClient();
    dbName = nanoid();

    await client.connect(dbName, 1, (db) => {
      db.createObjectStore(ObjectStore.Projects);
    });

    migration = new M20230816();
  });

  afterEach(() => {
    client.disconnect();
  });

  describe('Projects', () => {
    it('should do nothing', async () => {
      // Prepare
      const template = TestHelper.sampleProjectManifest();
      const docs: ProjectIDBEntry[] = [
        {
          version: 99,
          metadata: {
            ...template.metadata,
            id: 'test-project-1',
            name: 'Test project 1',
          },
          view: template.view,
          layerIds: ['layer-id-1', 'layer-id-2', 'layer-id-3'],
          layouts: {
            layoutIds: ['layout-id-1', 'layout-id-2', 'layout-id-3'],
            abcMapAttributionsEnabled: true,
          },
          sharedViews: {
            fullscreen: template.sharedViews.fullscreen,
            mapDimensions: template.sharedViews.mapDimensions,
            viewIds: ['shared-view-id-1', 'shared-view-id-2', 'shared-view-id-3'],
          },
        },
      ];

      await client.putAll(
        ObjectStore.Projects,
        docs.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<ProjectIDBEntry>(ObjectStore.Projects);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const template = TestHelper.sampleProjectManifest();
      const docs: ProjectIDBEntryV0[] = [
        {
          metadata: {
            ...template.metadata,
            id: 'test-project-1',
            name: 'Test project 1',
            containsCredentials: false,
          },
          view: template.view,
          layerIds: ['layer-id-1', 'layer-id-2', 'layer-id-3'],
          layoutIds: ['layout-id-1', 'layout-id-2', 'layout-id-3'],
          sharedViews: {
            fullscreen: template.sharedViews.fullscreen,
            mapDimensions: template.sharedViews.mapDimensions,
            viewIds: ['shared-view-id-1', 'shared-view-id-2', 'shared-view-id-3'],
          },
        },
      ];

      await client.putAll(
        ObjectStore.Projects,
        docs.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<ProjectIDBEntry>(ObjectStore.Projects);
      expect(entries).toMatchSnapshot();
    });
  });
});
