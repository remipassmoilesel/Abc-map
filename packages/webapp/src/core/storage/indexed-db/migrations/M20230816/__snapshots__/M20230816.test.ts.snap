// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M20230816 Projects should do nothing 1`] = `
Array [
  Object {
    "key": "test-project-1",
    "value": Object {
      "layerIds": Array [
        "layer-id-1",
        "layer-id-2",
        "layer-id-3",
      ],
      "layouts": Object {
        "abcMapAttributionsEnabled": true,
        "layoutIds": Array [
          "layout-id-1",
          "layout-id-2",
          "layout-id-3",
        ],
      },
      "metadata": Object {
        "id": "test-project-1",
        "name": "Test project 1",
        "public": false,
        "version": "1.3.0",
      },
      "sharedViews": Object {
        "fullscreen": false,
        "mapDimensions": Object {
          "height": 600,
          "width": 800,
        },
        "viewIds": Array [
          "shared-view-id-1",
          "shared-view-id-2",
          "shared-view-id-3",
        ],
      },
      "version": 99,
      "view": Object {
        "center": Array [
          1,
          2,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 1000,
        "rotation": 0,
      },
    },
  },
]
`;

exports[`M20230816 Projects should migrate 1`] = `
Array [
  Object {
    "key": "test-project-1",
    "value": Object {
      "layerIds": Array [
        "layer-id-1",
        "layer-id-2",
        "layer-id-3",
      ],
      "layouts": undefined,
      "metadata": Object {
        "id": "test-project-1",
        "name": "Test project 1",
        "public": false,
        "version": "1.3.0",
      },
      "sharedViews": Object {
        "fullscreen": false,
        "mapDimensions": Object {
          "height": 600,
          "width": 800,
        },
        "viewIds": Array [
          "shared-view-id-1",
          "shared-view-id-2",
          "shared-view-id-3",
        ],
      },
      "version": 3,
      "view": Object {
        "center": Array [
          1,
          2,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 1000,
        "rotation": 0,
      },
    },
  },
]
`;
